### ISSUE TITLE
enable gradle for extension development

### ISSUE THREAD

**Issue by ppalaga** [quarkusio/quarkus#3322]
A followup of #2113 . A Maven mojo was added via #3191 (see also #3321 )

**Comment by Dufgui** [comment]
If nobody start to work on it, i will try.

**Comment by maxandersen** [comment]
request makes sense but this would be post v1. I would actually prefer we had a native `qctl` or `qli` cli to do the bootstrapping. Its really backwards we have to tell users to install gradle or mvn ( + java) when we could just generate the initial project for them .... or even just give them a bit.ly url to curl  to get a zip from code.quarkus.io.

No need to strictly use either mvn or gradle for this step is my point ;)

**Comment by broxgit** [comment]
This would be largely beneficial for the expansion and adoption of Quarkus. We are incredibly interested in using Quarkus because it's a great framework and the extensions really open up possibilities for my team. However, our CI/CD architecture relies on Gradle (and many can argue that Gradle is superior) so being able to develop extensions with Gradle would be great!

**Comment by Dufgui** [comment]
sorry I have no more time to work on quarkus. But you can try a contribution. 
It's simple:
just like https://github.com/quarkusio/quarkus/blob/master/devtools/maven/src/main/java/io/quarkus/maven/CreateExtensionMojo.java

create a task here:
https://github.com/quarkusio/quarkus/tree/master/devtools/gradle/src/main/java/io/quarkus/gradle/tasks

**Comment by glefloch** [comment]
I think this one can be close as the `cli` will handle this right @maxandersen  ?

**Comment by maxandersen** [comment]
The cli Would be the place to have the command yes. Not sure if all is in place to develop a quarkus extension with gradle - @aloubyansky ?

**Comment by aloubyansky** [comment]
It's definitely possible to develop extensions with gradle. As to the tooling support, the maven extension-descriptor goal is pretty advanced now. But we have to start with the equivalent gradle support at some point. So, let's.

**Comment by glefloch** [comment]
I misunderstood the issue. I will start working on this next week then :)

**Comment by aloubyansky** [comment]
I didn't mean to add `createExtension` as a gradle task though. We would have a cli command for that but then we'd want some support from the gradle plugin in terms generating the proper extension metadata and descriptor.

**Comment by glefloch** [comment]
Yes, when I mention that the issue could be close as the cli is here I was thinking about the project generation part. But as you mentioned in your previous comment, the gradle plug-in don't do anything regarding the extension property file. That could be handled some how in the plug-in.

**Comment by maxandersen** [comment]
+1 I've updated issue title to reflect this. Have create extension cli command support gradle but for that have the remaining things done we still only have in maven ext plugin

**Comment by rdehuyss** [comment]
Would love to be able to add a real Quarkus extension in [JobRunr](https://github.com/jobrunr/jobrunr). JobRunr uses Gradle as build tool and I prefer not to mix this with maven.

**Comment by aloubyansky** [comment]
@rdehuyss do you mean you want to develop a Quarkus extension for the JobRunr? I can help you set it up in Gradle.

**Comment by rdehuyss** [comment]
@aloubyansky , that would be great!

**Comment by aloubyansky** [comment]
Are you on Zulip? That'll be easier to communicate.

**Comment by glefloch** [comment]
I can help too if needed.

**Comment by rdehuyss** [comment]
> Are you on Zulip? That'll be easier to communicate.

Not yet, I'll join in about an hour.

**Comment by aloubyansky** [comment]
Ping @aloubyansky and @glefloch when you're there.

**Comment by fooock** [comment]
what is the state of this? I saw in the devtool directory that there is a `gradle-extension-plugin` but I don't see any guide on how to use it. 

I'm trying to use the gradle `io.quarkus.extension` plugin in a multi-module setup, one of my subprojects has the dependency set in the build.gradle and I have the project structure created (deployment and runtime directories) but I see this in the log when gradle is configuring the project:

```
> Configure project :myproject-quarkus-extension
Unable to find deployment project with name: deployment. You can configure the deployment project name by setting the 'deploymentProject' property in the plugin extension.
```
Note that there is no `deploymentProject` property in the plugin [configuration](https://github.com/quarkusio/quarkus/blob/main/devtools/gradle/gradle-extension-plugin/src/main/java/io/quarkus/extension/gradle/QuarkusExtensionConfiguration.java#L9) maybe this should be [updated](https://github.com/quarkusio/quarkus/blob/main/devtools/gradle/gradle-extension-plugin/src/main/java/io/quarkus/extension/gradle/QuarkusExtensionPlugin.java#L87)

Do you know of any extension using Gradle?

**Comment by aloubyansky** [comment]
It's supposed to be https://quarkus.io/guides/writing-extensions#using-gradle

**Comment by glefloch** [comment]
Right there is a minimal quarkus extension gradle plugin. It has some limitation such as unit test support that need to be adresse before it can be fully used. 
Also, if you want examples, there are some gradle test samples that are using the plugin : https://github.com/quarkusio/quarkus/tree/main/integration-tests/gradle/src/main/resources/conditional-dependencies/ext-a

and also, this project have a gradle based quarkus extension but without using the plugin: https://github.com/jobrunr/jobrunr/tree/master/framework-support/jobrunr-quarkus-extension


Regarding your error, I will update at least the error message, thanks for reporting it.

**Comment by glefloch** [comment]
Right there is a minimal quarkus extension gradle plugin. It has some limitation such as unit test support that need to be adresse before it can be fully used.
Also, if you want examples, there are some gradle test samples that are using the plugin : https://github.com/quarkusio/quarkus/tree/main/integration-tests/gradle/src/main/resources/conditional-dependencies/ext-a

and also, this project have a gradle based quarkus extension but without using the plugin: https://github.com/jobrunr/jobrunr/tree/master/framework-support/jobrunr-quarkus-extension

Regarding your error, I will update at least the error message, thanks for reporting it.

**Comment by fooock** [comment]
thank you for pointing me in the right direction! 😀

**Comment by holly-cummins** [comment]
This is now available: https://quarkus.io/guides/building-my-first-extension/#gradle-setup

I don't know what PR added it, but I think it was a while ago
